pkgs/build-support/binary-cache/make-binary-cache.py

```diff
@@ -1,43 +1,78 @@
-
+from functools import partial
 import json
+from multiprocessing import Pool
 import os
 import subprocess
 
-with open(os.environ["NIX_ATTRS_JSON_FILE"], "r") as f:
-  closures = json.load(f)["closure"]
 
-os.chdir(os.environ["out"])
+def dropPrefix(path, nixPrefix):
+    return path[len(nixPrefix + "/") :]
+
+
+def processItem(item, nixPrefix, outDir):
+    os.chdir(outDir)
+
+    narInfoHash = dropPrefix(item["path"], nixPrefix).split("-")[0]
+
+    xzFile = "nar/" + narInfoHash + ".nar.xz"
+    with open(xzFile, "w") as f:
+        subprocess.run(
+            "nix-store --dump %s | xz -c" % item["path"],
+            stdout=f,
+            shell=True,
+            check=True,
+        )
+
+    fileHash = (
+        subprocess.run(
+            ["nix-hash", "--base32", "--type", "sha256", "--flat", xzFile],
+            capture_output=True,
+            check=True,
+        )
+        .stdout.decode()
+        .strip()
+    )
+    fileSize = os.path.getsize(xzFile)
+
+    finalXzFile = "nar/" + fileHash + ".nar.xz"
+    os.rename(xzFile, finalXzFile)
+
+    with open(narInfoHash + ".narinfo", "w") as f:
+        f.writelines(
+            (
+                x + "\n"
+                for x in [
+                    f"StorePath: {item['path']}",
+                    f"URL: {finalXzFile}",
+                    "Compression: xz",
+                    f"FileHash: sha256:{fileHash}",
+                    f"FileSize: {fileSize}",
+                    f"NarHash: {item['narHash']}",
+                    f"NarSize: {item['narSize']}",
+                    f"References: {' '.join(dropPrefix(ref, nixPrefix) for ref in item['references'])}",
+                ]
+            )
+        )
 
-nixPrefix = os.environ["NIX_STORE"] # Usually /nix/store
 
-with open("nix-cache-info", "w") as f:
-  f.write("StoreDir: " + nixPrefix + "\n")
+def main():
+    outDir = os.environ["out"]
+    nixPrefix = os.environ["NIX_STORE"]
+    numWorkers = int(os.environ.get("NIX_BUILD_CORES", "4"))
 
-def dropPrefix(path):
-  return path[len(nixPrefix + "/"):]
+    with open(os.environ["NIX_ATTRS_JSON_FILE"], "r") as f:
+        closures = json.load(f)["closure"]
 
-for item in closures:
-  narInfoHash = dropPrefix(item["path"]).split("-")[0]
+    os.chdir(outDir)
+    os.makedirs("nar", exist_ok=True)
 
-  xzFile = "nar/" + narInfoHash + ".nar.xz"
-  with open(xzFile, "w") as f:
-    subprocess.run("nix-store --dump %s | xz -c" % item["path"], stdout=f, shell=True)
+    with open("nix-cache-info", "w") as f:
+        f.write("StoreDir: " + nixPrefix + "\n")
 
-  fileHash = subprocess.run(["nix-hash", "--base32", "--type", "sha256", "--flat", xzFile], capture_output=True).stdout.decode().strip()
-  fileSize = os.path.getsize(xzFile)
+    with Pool(processes=numWorkers) as pool:
+        worker = partial(processItem, nixPrefix=nixPrefix, outDir=outDir)
+        pool.map(worker, closures)
```

Comment:
If we did the strings-for-errors, it would look like this.

```suggestion
        results = pool.map(worker, closures)
        for result in results:
           if result is not None:
               print(result, file=sys.stderr, flush=True)
        return 0 if all(r is None for r in results) else 1
```